### ISSUE TITLE
Issues with ACES AP0 exports:

### ISSUE THREAD

**Issue by avclubvids** [Beep6581/RawTherapee#6732]
**Short description**
I'm trying to figure out the recommended path to exporting linear gamma ACES AP0 images from a DNG in RawTherapee. I have been using the below settings in RT, and I have tried saving 16-bit, 16-bit float, and 32-bit float TIFFs and 16-bit PNG. The resulting images all work fine in OSX (Finder/Preview), but I get the following results with different software:

- In Adobe (Lightroom, Photoshop) the 32 and 16 float TIFFs are lifted, the 16-bit TIFF and PNG are correct
- In Resolve/Fusion the 16-bit float TIFF shows up as 16-bit int with terribly wrong color, and the other three show up less saturated than expected when interpreted as ACES AP0 with linear gamma
- In Affinity Photo the 16-bit float is broken like Resolve/Fusion, I cannot determine if it is interpreting it as 16-bit int or float, the other three appear as expected

I have not tested other software yet, but I can. Screenshots below showing the issues.

<img width="1506" alt="PhotoshopComparisonScreenshot" src="https://user-images.githubusercontent.com/9397841/228984542-04361e38-7894-404c-ac6f-39683fcf2c8a.png">

<img width="922" alt="FusionComparisonScreenshot" src="https://user-images.githubusercontent.com/9397841/228984552-9eb1f8ab-c7e1-4594-a9db-2a74fe81e7e1.png">

<img width="1792" alt="AffinityComparisonScreenshot" src="https://user-images.githubusercontent.com/9397841/228984562-f3b1b9af-716c-4e8f-9541-5fb4486bc80e.png">

**Steps to reproduce**
1. Under Color Management:
2. Input Profile -> Camera standard
3. Working Profile -> ACESp0
4. Abstract Profile -> Linear g=1
5. Destination primaries: ACESp0
6. Output Profile -> RTv4_ACES-AP0
7. Put to queue (TIFF 16-bit, TIFF 16-bit float, TIFF 32-bit float, PNG 16-bit)

**Expected behavior**
I would hope that the exported images would look identical in all software, and the color would be exactly right when interpreted as ACES AP0.

**Additional information**
- RawTherapee v5.9 (installed in the last 48hrs, so it is latest)
- OSX Ventura 13.1

Other useful information:
- Image files + sidecars (pp3):
[RT_ACES_Exports_tiff.zip](https://github.com/Beep6581/RawTherapee/files/11116854/RT_ACES_Exports_tiff.zip)
[RT_ACES_Exports_png.zip](https://github.com/Beep6581/RawTherapee/files/11116855/RT_ACES_Exports_png.zip)

**Comment by kmilos** [comment]
No problem w/ these images in darktable, so it's likely that color management (and/or float TIFF reading) in all these other apps is broken.

![image](https://user-images.githubusercontent.com/4973094/229149588-bf844472-0b76-4d28-9294-9ce4e196ef0a.png)

Fine in GIMP 2.99 as well.

![image](https://user-images.githubusercontent.com/4973094/229151754-9e3ad147-0326-4117-bc25-a72eaaa44802.png)

**Comment by TechXavAL** [comment]
First things first: bit depth has nothing to do with color management. Bit depth only has to do with the reproducible number of steps between black and white (or the corresponding min and max values of each color channel, rgb). Integer numbers are more precise, but they have _less steps_. On the other hand, floating numbers are less precise (although with 32bit calculations, the error is negligible) and have _more steps_.

But even though you can choose whichever fits best for you workflow, you must ensure the next program to use will read such bit depth (and not all programs do it, or do it well).

Now to color management in RT: no matter what you do in RT, if you use _RTv4_ACES-AP0_ it will be encoded with a TRC (the so called _gamma_) like the one present in sRGB.

In other words, the output image **will not be linearly encoded**. **It won't have a _gamma_=1.0**. It will have the sRGB TRC.

Thus, if you load such image into a program and you tell it that is a linear image, then you won't get good results.

dt and gimp know how that image is encoded (with an sRGB TRC), and with their color management engines _know what to do_ to show a good image.

Workaround: download and use the appropriate [ACESp0 linear profile from Elle Stone (ACES-elle-V4-g10.icc)](https://github.com/ellelstone/elles_icc_profiles/tree/master/profiles) as output profile, or create and use your own linear profile with the included ICC Profile Creator.

**Comment by kmilos** [comment]
> First things first: bit depth has nothing to do with color management.

One would think, right? But it's not just that Adobe example, I think also Windows Photo viewer treats those float TIFFs differently to integer ones. So, it looks like, when they see float they think they have the right to discard the file embedded profile TRC and assume a linear one. So in a sense, yeah, it is "safer" to export with a linear one.

In any case, I don't think there is anything wrong per se w/ RT and these files. Maybe one should not ever have a TRC w/ an AP0 profile, as that is the expectation/specification?

**Comment by avclubvids** [comment]
Ok, I downloaded and used [ACESp0 linear profile from Elle Stone (ACES-elle-V4-g10.icc)](https://github.com/ellelstone/elles_icc_profiles/tree/master/profiles) – it seems to have worked as desired. Thank you very much @TechXavAL. The great news is that color is identical in every application that I have tried so far, which is fantastic. The location that is documented for OSX for these profiles is not correct, I had to open the application container and put them inside it with the ones that ship with RT, let me know if you'd like that path and method detailed in order to update the documentation?

I am not familiar with TRCs so I will look them up before saying anything about them, but in general, anyone going to ACES AP0 would not want to be using sRGB primaries or have a non-linear gamma applied. ACES AP0 is supposed to have linear gamma; AP1 can use linear, log, or other gamma values. So even though I have not done the research yet, I suspect @kmilos is right that TRCs should not be applied to AP0 outputs no matter what.

The bad news is that there seems to be a separate issue (which is probably not ACES related) with RT's 16-bit float TIFFs.  I will report it in a new separate issue from this one. My guess is that RT, Darktable, and Gimp probably all use the same or very similar open-source TIFF decoding library, thus they all show the same as each other. But non-open source apps seem to take issue with this encoding. You can see the issue is present in Affinity, Fusion, and Nuke. 

Here are screenshots and new ACES AP0 outputs for anyone who wants to play along at home:

<img width="1134" alt="NukeACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229269638-c7ffefeb-45a8-4dbb-9603-1aaf980c3590.png">
<img width="1017" alt="PreviewACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229269626-741b11f8-0f4e-451b-a3f6-6f64c564d8c9.png">
<img width="1064" alt="PhotoshopACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229269630-35c01a89-7493-41b3-bb1c-8c1ca3f21808.png">
<img width="1792" alt="AffinityACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229269632-0a221fd0-123c-4d33-99d2-6fa8edf969e2.png">
<img width="1326" alt="FusionACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229269636-5e51d836-8113-488c-87b9-23a8f4d43a07.png">

[16bit_tiffs.zip](https://github.com/Beep6581/RawTherapee/files/11128618/16bit_tiffs.zip)
[32bit_tiff.zip](https://github.com/Beep6581/RawTherapee/files/11128619/32bit_tiff.zip)
[png_v2.zip](https://github.com/Beep6581/RawTherapee/files/11128620/png_v2.zip)

**Comment by avclubvids** [comment]
Actually, one thing I'm seeing after I posted the above screenshots is that Nuke and Fusion [the most "professional" of the bunch in terms of adhering to ACES specs] are less saturated than Photoshop, Affinity, and Preview. The less saturated look is what is desired, so there is something happening in how the other apps are reading the images. It's not necessarily an RT encoding issue, it's probably an ACES decoding issue on the part of Preview, Adobe, and Affinity.

**Comment by kmilos** [comment]
> The bad news is that there seems to be a separate issue (which is probably not ACES related) with RT's 16-bit float TIFFs. I will report it in a new separate issue from this one.

I really think there is nothing to report here, and you should take this up with the other apps.

dt uses the reference [Imath](https://github.com/AcademySoftwareFoundation/Imath) library for this, authored by the 'inventors' of 16b float.

And why do you assume that, something being open source makes it of 'lesser' quality?

**Comment by avclubvids** [comment]
Please read my comments more carefully. I referred to Nuke and Fusion as more "professional" than Affinity, Adobe, and Preview, and only as pertaining to ACES specification adherence. I said nothing about open source being of lesser quality.

This is not about closed vs. open source, it is about making sure that everything is reading and writing the same files the same way. Ours is a large industry with hundreds of individual pieces of software, and standards are important. If there are programs that read or write data differently than many others, that's an issue. RT is writing 16-bit float TIFFs in an odd manner, one that is breaking in programs used by millions of end users. As you pointed out yourself, Microsoft Photo viewer is having an issue with those TIFFs too, that's a lot of users.

**Comment by kmilos** [comment]
> RT is writing 16-bit float TIFFs in an odd manner

Again, why do you assume RT is doing something 'odd' here, and not these 'professional' apps?

**Comment by kmilos** [comment]
I have double-checked your `16bit_tiffs.zip` again, and there is absolutely nothing 'odd' about the 16-bit float file w.r.t. to the TIFF specification.

**Comment by kmilos** [comment]
> As you pointed out yourself, Microsoft Photo viewer is having an issue with those TIFFs too

And, as I mentioned above, it's not about not being able to read the files, it was about color (mis)management in the case of floats.

> that's a lot of users

It is not. People using 16-bit float are (sadly) a statistical error compared to mainstream 8/16-bit integer use cases (for TIFF; people using EXRs are also not that many). Hence the lack of robust testing and implementation.

**Comment by Lawrence37** [comment]
ImageMagick also uses libtiff, but it is well-tested and widely used. It is able to correctly read the 16-bit floating point TIFF files from RawTherapee. How do we know which program(s) are not handling the format incorrectly?

Since you have access to those proprietary software, can you check the open-source licenses to verify they are not using libtiff?

**Comment by avclubvids** [comment]
A few applications in the VFX space use [Open Image IO](https://sites.google.com/site/openimageio/home) (OIIO), I'll do a little digging and see what other libraries that uses. There is also an open-source initiative by the Academy Software Foundation called [rawtoaces](https://github.com/AcademySoftwareFoundation/rawtoaces) that aims to provide a scriptable path from popular raw formats to ACES, but it is pretty early stage still. Also worth mentioning that [Open Color IO](https://opencolorio.org/) (OCIO) has a lot of tools for converting into and out of different gamuts/gammas, and many of the large tools in VFX use or at least support the use of those tools, which provides a lot of consistency between applications. 

I tested further on Blender, Cinesync Play, DJV View, PTGUI, Safari, and Topaz AI:

Blender same as Nuke/Fusion/Affinity:
<img width="740" alt="BlenderACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229583864-bf899f4c-9598-4b51-84bf-07f58cd9e11b.png">

Cinesync & DJV View same color as Nuke/Fusion but 16-float tiff works (the same developer works on both so I would expect them to be the same):
<img width="1702" alt="CinesyncACESCompare_tiff32vsPNG_v002" src="https://user-images.githubusercontent.com/9397841/229584231-5333ad1b-6656-4745-8e66-dc818e6d1af9.png">
<img width="1702" alt="CinesyncACESCompare_tiffs_v002" src="https://user-images.githubusercontent.com/9397841/229584258-babf65e3-8d4f-4451-b4a8-c0e4791985c1.png">
<img width="1198" alt="DJVViewACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229584270-181df72f-5cd6-40f6-afe2-ecc56d0c0f60.png">

PTGui does not support 16-float TIFF:
<img width="1904" alt="PTGUIACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229584430-88f107af-f6ab-4870-b304-4bdc49dd4c9a.png">

Safari and Topaz same as Preview (more saturated than desired but 16-float works):
<img width="1582" alt="SafariACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229584766-86f430e6-cf6f-46f1-800c-e43555c3a390.png">
<img width="1904" alt="TopazACESCompare_v002" src="https://user-images.githubusercontent.com/9397841/229584796-5799a3fd-df31-4542-b7a6-2078e3b7649c.png">

I don't have any other applications installed that read or write images, but I'll see if I can ask someone to test with Silhouette, Maya, Houdini, and C4D.

**Comment by Lawrence37** [comment]
OpenImageIO uses libtiff too. There is no indication of any issues on the RawTherapee side, therefore closing this issue.
